design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,291 @@
+but may become tricky to evolve and we need a good API to do this, especially if it is necessary to
+specify complicated nested structs. Below are three candidate ways to specify the schema:
+
+```java
+@UdfSchema
+public static final Schema SCHEMA = SchemaBuilder.struct()...build();
+
+@Udf("Checks if the employee has a valid record (i.e. contains a valid name and email)")
+@UdfReturn(value  = "STRUCT<'A' INT, 'B' VARCHAR>") // sample specification annotation
+@UdfReturn(file   = 'schema_def.kschema')           // another way pointing to a file
+@UdfReturn(schema = "name.space.MyClass.SCHEMA")    // another way that would resolve a java object
+public Struct generate() {
+  return new Struct(...);
+}
+```
+
+### Generics
+
+Today, we eagerly convert UDF output types to corresponding Connect schema types at the time that we
+compile the UDFs and not the time that we use them in a `SELECT` clause. Since there is no 
+corresponding "unresolved" type in Connect, we fail this conversion. To solve this problem, we can
+infer the type from the source schema and compile the code to work with any component type (e.g.
+omit the generics from the compiled code).
+
+```java
+@Udf("returns a sublist of 'list'")
+public <T> List<T> sublist(
+    @UdfParameter final List<T> list, 
+    @UdfParameter final int start, 
+    @UdfParameter final int end) {
+  return list.subList(start, end);
+}
+```
+
+We need to ensure that the output type can be inferred. This means that we need to fail if the 
+signature is something like `<T> List<T> convert(List<String> list)`. This can be done during 
+compilation by following a simple rule: any generic type used in the output must be present in at
+least one of the parameters. We can access this information via reflection.
+
+**NOTE:** Supporting a wildcard type (e.g. `<?>`) is not covered by this design since we would not
+be able to generate the output schema of select statements.
+
+### Varargs
+
+Varargs boils down to supporting native Java arrays as parameters and ensuring that component types 
+are resolved properly. Anytime a method is registered with variable args (e.g. `String... args`) we 
+will register a corresponding function with the wrapping array type (e.g. `String[]`). At runtime, 
+if no function matches the explicit parameter types, we will fallback to search for any array-type 
+parameter method declaration of the same type iff all parameters are of a single type. If any
+parameter is null, it will be considered valid for any vararg declaration.
+
+This proposal does not cover supporting `Object` as a parameter to a method in order to allow for 
+"generic variable argument" UDFs, but supporting it can be an extension.
+
+```java
+@Udf("returns a sublist of 'list' t")
+public int sum(@UdfParameter final int... args) {
+  return Arrays.stream(args).sum();
+}
+```
+
+### Complex Aggregation
+
+We will allow users to supply an additional method `VR export(A agg)`. This method will be taken
```

Comment:
Out of curiosity, what does `VR` stand for? I gather it's the return type but can't piece together the acronym.

Reply:
I assume it stands for `ValueReturn`. I took it from KStreams:
```
    <VR> KStream<K, VR> mapValues(final ValueMapper<? super V, ? extends VR> mapper);
```